add fuse admm/matmul back to linear pass

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58 add fuse admm/matmul back to linear pass**
* #57 [codemod] deprecated macros

wanchaol committed:
add fuse admm/matmul back to linear pass

wanchaol committed:
Update on "add fuse admm/matmul back to linear pass"

add fuse admm/matmul back to linear pass

gh-metadata: pytorch tvm 58 gh/wanchaol/4/head

bwasti commented on torch_tvm/fuse_linear.cpp:
```diff
@@ -0,0 +1,50 @@
+    graph(%input, %weight, %bias, %4):
+        %weight_t = aten::t(%weight)
+        %output = aten::matmul(%input, %weight_t)
+        %res = aten::add_(%output, %bias, %4)
```
should we also match on non-inplace add?

wanchaol replied:
We can but I don't think we need, people using `nn.Linear` will always get this inplace `add` through tracing/scripting, outplace `add` will be matched if people manually write `matmul + outplace add`.

bwasti commented on test/test_operators.py:
```diff
@@ -206,9 +206,15 @@ def test_linear(self, shape, out_features):
         def linear(input, weight, bias):
             return F.linear(input, weight, bias) + 2.0
 
+        def linear_no_bias(input, weight):
```
maybe we should test the pattern matching in python?

wanchaol replied:
Yeah sure we will need to pybind the pass to python, I will add a test

bwasti replied:
can we avoid pybind by manually constructing a torchscript model containing t + matmul + _add and just ensuring "linear" exists in the graph_for call?

bwasti reviewed (approved):
nice! :)

wanchaol committed:
address comments on "add fuse admm/matmul back to linear pass"

add fuse admm/matmul back to linear pass

gh-metadata: pytorch tvm 58 gh/wanchaol/4/head
